WIP: Huawei ap5030dn

CodingMarco opened the pull request:
Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

CodingMarco committed:
Add initial Huawei ap5030dn model

CodingMarco committed:
Flash layout from uboot cmdline

CodingMarco committed:
Remove unneeded USB driver, move squashfs size to property

CodingMarco committed:
dts: Add 9600 baudrate (same as u-boot) & watchdog

CodingMarco committed:
Make firmware partition as large as vendor firmware has it

CodingMarco committed:
Correct flash partitions, add wifi & ethernet MACs and PHYs

CodingMarco committed:
Use mtd concat to combine SysImageA & SysImageB vendor partitions into firmware partition

CodingMarco committed:
generic.mk: generate kernel and squashfs

CodingMarco committed:
Device tree: move art locations to separate referencing node, add eth0 MAC address

CodingMarco committed:
generic.mk: Pad rootfs to minimum required size, otherwise u-boot will not flash it

CodingMarco committed:
02_network: Set both ethernet interfaces as LAN (not WAN) interfaces

CodingMarco committed:
generik.mk: Make ROOTFS_MIN_SIZE larger because u-boot otherwise gives 'bad magic number' error after flashing

CodingMarco committed:
Make eth0 work using fixed-link

CodingMarco committed:
02_network: Add AP5030DN to lan_mac and label_mac setup

CodingMarco committed:
dts: Remove eth1 MAC address increment since vendor FW also doesn't have it

CodingMarco committed:
dts: Remove unnecessary wlan0 section from pcie0

CodingMarco committed:
dts: Remove phy0 switch register values since there is no such switch

CodingMarco committed:
dts: Add LEDs, increase watchdog margin to 1s (datasheet says 1.6s)

CodingMarco committed:
dts: Add 'Default' GPIO-button as restart button

CodingMarco committed:
Add AP5030DN u-boot env-tools support

CodingMarco committed:
dts: Change u-boot env label to OpenWrt-default 'u-boot-env' instead of Huawei-custom 'BootEnv'

CodingMarco committed:
dts: Add correct phy addresses & pll-data for 1000Mbps (100/10 still not working)

CodingMarco committed:
dts: pll-data values left-over from u-boot

CodingMarco committed:
Use fixed-link for eth0 again

CodingMarco committed:
dts: Decrease watchdog margin to 100ms

CodingMarco committed:
dts: Rename BiosA / BiosB to more openwrt-compatible names u-boot-a and u-boot-b

CodingMarco committed:
dts: Make BootupA/B writable as they contain kernel start address which has to be set for sysupgrade

CodingMarco committed:
dts: Merge fwconcat0/1 with ConfigA/B to gain more usable flash space

CodingMarco committed:
Use phy0 for eth0 again since it doesn't make a difference. U-Boot still seems to do some PHY initialization magic when auto-boot is interrupted

CodingMarco committed:
sysupgrade platform.sh: Add Huawei AP5030DN: Write beginning of 'firmware' partition to BootupA/B as boot address.  doesn't seem to work, so use dd

CodingMarco committed:
Remove flashAll partition which was only used for debugging purposes

CodingMarco committed:
Fix MAC addresses. Use offsets 0-3 for eth0, eth1, wlan0 and wlan1 respectively

CodingMarco committed:
Clean up device tree

CodingMarco committed:
Add kernel 5.15 and 6.1 patches to add suspend & resume callbacks to BCM54612E driver

CodingMarco committed:
Move device makefile definition down to keep alphabetical order

CodingMarco committed:
Compress ramdisk with LZMA for AP5030DN

Othwise, the boot will fail.

CodingMarco committed:
Remove 5.15 patch again since it is not needed anymore

CodingMarco committed:
Add patch to backport-6.1 since it is now upstream in 6.8

CodingMarco committed:
Use nvmem fixed-layout for mac address cal art

CodingMarco committed:
Initial draft for fixing LZMA & watchdog problems using lzma-loader

blocktrron commented on target/linux/ath79/generic/base-files/lib/upgrade/platform.sh:
```diff
@@ -67,6 +67,13 @@ platform_do_upgrade() {
 		ROOTFS_FILE="root.squashfs"
 		platform_do_upgrade_failsafe_datachk "$1"
 		;;
+	huawei,ap5030dn)
+		# Write beginning address of "firmware" partition as KernelA address and KernelB address, each to BootupA & BootupB
+		echo "Setting Huawei-custom kernel addresses..."
+		echo -n -e "\x9e\x10\x00\x00\x9e\x10\x00\x00" | dd of=/dev/mtdblock1 bs=1 seek=$((0x254)) conv=notrunc
+		echo -n -e "\x9e\x10\x00\x00\x9e\x10\x00\x00" | dd of=/dev/mtdblock2 bs=1 seek=$((0x254)) conv=notrunc
+		default_do_upgrade "$1"
+		;;
```
Is this the offset the Huawei bootloader uses to determine the firmware offset to boot from? You can use mtd lookup by name here nevertheless.

blocktrron replied:
Also drop the echo. I assume this is a debug-leftover

CodingMarco replied:
Yes it is. Added it to the comment

blocktrron commented on target/linux/ath79/image/Makefile:
```diff
@@ -47,6 +47,12 @@ define Build/append-loader-okli-uimage
 	cat "$(KDIR)/loader-$(word 1,$(1)).uImage" >> "$@"
 endef
 
+define Build/prepend-loader-okli
+	cat "$(KDIR)/loader-$(word 1,$(1)).$(LOADER_TYPE)" > "$@.new"
+	cat "$@" >> "$@.new"
+	mv "$@.new" "$@"
+endef
+
```
Short description of arguments / How to use

CodingMarco replied:
Removed it in the end since it is equivalent to loader-okli with bs=1

blocktrron commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1813,6 +1813,26 @@ define Device/hiwifi_hc6361
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct kmod-ath10k-ct
+  LOADER_TYPE := bin
+  LOADER_FLASH_OFFS := 0x110040
+  IMAGE_SIZE := 7680k
+  ROOTFS_MIN_SIZE := 12240k
```
This is weird, image-size < rootfs-min-size?

CodingMarco replied:
Removed this development leftover.

blocktrron commented on target/linux/generic/backport-6.1/738-net-phy-bcm54612e-add-suspend-resume.patch:
```diff
@@ -0,0 +1,27 @@
+From 380b50ae3a04222334a3779b3787eba844b1177f Mon Sep 17 00:00:00 2001
```
If this is a backport, prepend the kernel version this commit first appears in. (Filename)

```
738-vx.y-net-phy-bcm54612e-add-suspend-resume.patch
```

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/dts/qca9550_huawei_ap5030dn.dts:
```diff
@@ -0,0 +1,230 @@
+	nvmem-cells = <&macaddr_art_2005b 0>;
+	nvmem-cell-names = "mac-address";
+
+	pll-data = <0xa6000000 0x80000101 0x80001313>;
```
Where are these from? Incorrect timings can lead to issues on longer cable runs, so this either needs to be verified by hardware or via statistics (Floodping at all link-speeds is good enough)

blocktrron commented on target/linux/ath79/dts/qca9550_huawei_ap5030dn.dts:
```diff
@@ -0,0 +1,230 @@
+				read-only;
+			};
+
+			partition@80000 {
+				label = "BootupA";
+				reg = <0x80000 0x20000>;
+			};
+
+			partition@a0000 {
+				label = "BootupB";
+				reg = <0xa0000 0x20000>;
+			};
```
Adding a short comment about the usage of those in the upgrade-process would be a nice touch.

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/dts/qca9550_huawei_ap5030dn.dts:
```diff
@@ -0,0 +1,230 @@
+		compatible = "gpio-keys";
+
+		restart {
+			label = "Restart button";
```
`label = "reset";`

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/dts/qca9550_huawei_ap5030dn.dts:
```diff
@@ -0,0 +1,230 @@
+		};
+	};
+
+	watchdog {
+		compatible = "linux,wdt-gpio";
+		gpios = <&gpio 15 GPIO_ACTIVE_HIGH>;
+		hw_algo = "toggle";
+		hw_margin_ms = <100>;
+		always-running;
+	};
+
```
Pinctrl to reconfigure the pin-function to reflect the usage as sw-watchdog toggle.

CodingMarco replied:
done, also had to disable SoC-internal watchdog

blocktrron commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1813,6 +1813,26 @@ define Device/hiwifi_hc6361
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | lzma | loader-kernel | uImage none
+  IMAGE_SIZE := 30720k
+  IMAGES += kernel.bin squashfs.bin
+  IMAGE/kernel.bin := | append-kernel
```
The pipe looks incorrect

CodingMarco replied:
Removed this development leftover.

blocktrron commented on target/linux/ath79/image/lzma-loader/src/board.c:
```diff
@@ -182,8 +182,38 @@ static inline void mr18_init(void)
+	const unsigned int ap5030dn_watchdog_gpio = 15;
+	unsigned int gpiobase, spibase, reg;
+
+	spibase = KSEG1ADDR(AR71XX_SPI_BASE);
```
I dont see this used

CodingMarco replied:
Removed this development leftover.

blocktrron commented on target/linux/ath79/image/lzma-loader/src/board.c:
```diff
@@ -182,8 +182,38 @@ static inline void mr18_init(void)
+		reg & ~(1 << ap5030dn_watchdog_gpio));
+
+	// Set MUX to output CPU_CLK/4 on GPIO15
+	reg = READREG(gpiobase + AR934X_GPIO_REG_FUNC);
+	WRITEREG(gpiobase + AR934X_GPIO_REG_FUNC,
+		reg | (1 << 7));
```
Comment is misleading, enables observation of the clock for the GPIO subsys, does not designate the GPIO pin

CodingMarco replied:
fixed

blocktrron commented on target/linux/ath79/image/lzma-loader/src/board.c:
```diff
@@ -182,8 +182,38 @@ static inline void mr18_init(void)
+		reg | (1 << 7));
+
+	// Output stuff on GPIO15 by setting bit 31:24 in gpiobase + AR934X_GPIO_REG_OUT_FUNC3
+	reg = READREG(gpiobase + AR934X_GPIO_REG_OUT_FUNC3);
+	WRITEREG(gpiobase + AR934X_GPIO_REG_OUT_FUNC3,
+		reg | (84 << 24));
```
I'd drop the register calculation from the comment. 

Also convert the `84` to hex

CodingMarco replied:
done. Is the introduced define ok? Where to put it? Which SoC prefix to use?

blocktrron replied:
Looks good

CodingMarco committed:
Improve watchdog-related init in LZMA loader

CodingMarco committed:
Add pinctrl to watchdog

... but now saying "a legacy watchdog module is probably present"
instead of "ath79-wdt 18060008.wdt: unable to register misc device,
err=-16"

CodingMarco committed:
Label restart button 'reset'

CodingMarco committed:
Add comment about the BootupA/B partitions

CodingMarco committed:
Disable SoC-internal watchdog to avoid clash with driver of external one

CodingMarco committed:
generik.mk: Remove leftovers from development/debugging

CodingMarco committed:
platform.sh: Use mtd lookup by name

CodingMarco committed:
Remove prepend-loader-okli again since it does essentially the same as loader-okli, just always with a block size of 1

CodingMarco committed:
Add description comment to Build/loader-okli

CodingMarco committed:
Rename kernel patch to include version of first release

CodingMarco committed:
platform.sh: Improve explanation of kernel start address

blocktrron commented on target/linux/ath79/dts/qca9550_huawei_ap5030dn.dts:
```diff
@@ -0,0 +1,247 @@
+	leds {
+		compatible = "gpio-leds";
+
+		led_power_green: power_green {
+			label = "green:power";
+			gpios = <&gpio 18 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		led_power_red: power_red {
+			label = "red:power";
+			gpios = <&gpio 11 GPIO_ACTIVE_HIGH>;
+		};
```
LED functions

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/image/Makefile:
```diff
@@ -33,6 +33,7 @@ define Build/loader-okli-compile
 endef
 
 # Arguments: <output name> <kernel offset>
+# Pads the specified loader to a multiple of <kernel offset> and prepends the result
```
Remove this

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1813,6 +1813,21 @@ define Device/hiwifi_hc6361
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct kmod-ath10k-ct
+  LOADER_TYPE := bin
+  LOADER_FLASH_OFFS := 0x111E00
+  IMAGE_SIZE := 7680k
```
30720k

CodingMarco replied:
done

blocktrron commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1813,6 +1813,21 @@ define Device/hiwifi_hc6361
+  DEVICE_MODEL := AP5030DN
+  DEVICE_PACKAGES := ath10k-firmware-qca988x-ct kmod-ath10k-ct
+  LOADER_TYPE := bin
+  LOADER_FLASH_OFFS := 0x111E00
```
Add a `KERNEL_SIZE` parameter the size of the first partition so that the kernel does not get split between both partitions.

CodingMarco replied:
done

CodingMarco committed:
Fix eth0 pll for all speeds, modern LED syntax

CodingMarco committed:
Fix IMAGE_SIZE, add KERNEL_SIZE

CodingMarco committed:
Remove loader-okli comment

CodingMarco committed:
Fix LOADER_FLASH_OFFS

CodingMarco committed:
Add backport-6.6 patch
